cqrs-codegen/src/event/registered_event.rs

```diff
@@ -0,0 +1,97 @@
+//! Codegen for [`cqrs::RegisteredEvent`].
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::Result;
+use synstructure::Structure;
+
+use crate::util;
+
+/// Name of the derived trait.
+const TRAIT_NAME: &str = "RegisteredEvent";
+
+/// Implements [`crate::derive_registered_event`] macro expansion.
+pub fn derive(input: syn::DeriveInput) -> Result<TokenStream> {
+    util::derive(input, TRAIT_NAME, derive_struct, derive_enum)
+}
+
+/// Implements [`crate::derive_registered_event`] macro expansion for structs.
+fn derive_struct(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    let body = quote! {
+        #[inline(always)]
+        fn type_id(&self) -> ::core::any::TypeId {
+            ::core::any::TypeId::of::<Self>()
+        }
+    };
+
+    super::render_struct(&input, quote!(::cqrs::RegisteredEvent), body, None)
+}
+
+/// Implements [`crate::derive_registered_event`] macro expansion for enums
+/// via [`synstructure`].
+fn derive_enum(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    util::assert_attr_does_not_exist(&input.attrs, super::ATTR_NAME)?;
```

Comment:
This derive macro doesn't interact with `#[event]` attributes, so it shouldn't validate them. It's not responsible for that.